<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Guided-Capstone-Step-4---Pre-processing-and-Training-Data-Development" data-toc-modified-id="Guided-Capstone-Step-4---Pre-processing-and-Training-Data-Development-0"><span class="toc-item-num">0&nbsp;&nbsp;</span>Guided Capstone Step 4 - Pre-processing and Training Data Development</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#The-Data-Science-Method" data-toc-modified-id="The-Data-Science-Method-0.0.1"><span class="toc-item-num">0.0.1&nbsp;&nbsp;</span><strong>The Data Science Method</strong></a></span></li></ul></li><li><span><a href="#Create-dummy-features-for-categorical-variables" data-toc-modified-id="Create-dummy-features-for-categorical-variables-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Create dummy features for categorical variables</a></span></li><li><span><a href="#Standardize-the-magnitude-of-numeric-features" data-toc-modified-id="Standardize-the-magnitude-of-numeric-features-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Standardize the magnitude of numeric features</a></span></li><li><span><a href="#Split-into-training-and-testing-datasets" data-toc-modified-id="Split-into-training-and-testing-datasets-0.3"><span class="toc-item-num">0.3&nbsp;&nbsp;</span>Split into training and testing datasets</a></span></li></ul></li><li><span><a href="#Guided-Capstone-Step-5---Modeling" data-toc-modified-id="Guided-Capstone-Step-5---Modeling-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Guided Capstone Step 5 - Modeling</a></span><ul class="toc-item"><li><span><a href="#Fit-Models-with-Training-Data-Set" data-toc-modified-id="Fit-Models-with-Training-Data-Set-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Fit Models with Training Data Set</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Model-1" data-toc-modified-id="Model-1-1.1.0.1"><span class="toc-item-num">1.1.0.1&nbsp;&nbsp;</span>Model 1</a></span></li></ul></li></ul></li><li><span><a href="#Review-Model-Outcomes-—-Iterate-over-additional-models-as-needed." data-toc-modified-id="Review-Model-Outcomes-—-Iterate-over-additional-models-as-needed.-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Review Model Outcomes — Iterate over additional models as needed.</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Model-2" data-toc-modified-id="Model-2-1.2.0.1"><span class="toc-item-num">1.2.0.1&nbsp;&nbsp;</span>Model 2</a></span></li><li><span><a href="#Model-3" data-toc-modified-id="Model-3-1.2.0.2"><span class="toc-item-num">1.2.0.2&nbsp;&nbsp;</span>Model 3</a></span></li></ul></li></ul></li><li><span><a href="#Identify-the-Final-Model" data-toc-modified-id="Identify-the-Final-Model-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Identify the Final Model</a></span></li></ul></li></ul></div>

# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [135]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
from sklearn import metrics
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [136]:
df = pd.read_csv('data/step3_output.csv')

                            Name       state  summit_elev  vertical_drop  \
0                 Alyeska Resort      Alaska         3939           2500   
1               Hilltop Ski Area      Alaska         2090            294   
2               Arizona Snowbowl     Arizona        11500           2300   
3            Sunrise Park Resort     Arizona        11100           1800   
4  Yosemite Ski & Snowboard Area  California         7800            600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      1        0.0          0          2     2       0  ...   
1      0        0.0          0          0     0       1  ...   
2      0        0.0          1          0     2       2  ...   
3      0        0.0          0          1     2       3  ...   
4      0        0.0          0          0     0       1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0      113.000000             150.0       60.0   
1               30.0

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [137]:
# Remove state column and create dummies variables
df = pd.concat([df.drop('state', axis=1), pd.get_dummies(df.state)], axis=1)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be droppped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [138]:
# Remove column Name from the dataframe
df.drop('Name', axis=1, inplace=True)

In [139]:
# Assign Y to AdultWeekend and the rest of the dataframe to X
Y = df['AdultWeekend']
X = df.drop('AdultWeekend', axis=1)

In [140]:
# Standardize the scale of each features in X
# scaler = StandardScaler().fit(X)
# X_scaled=scaler.transform(X)

# Using scale() scales the entire dataframe while the StandardScaler() scales each column. 
X_scaled = scale(X, axis=0, with_mean=True, with_std=True, copy=True)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [141]:
# Create a 75/25 train/test split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, Y, test_size=0.25, random_state=42)

Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [142]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [143]:
# Create a prediction y_pred
y_pred = lm.predict(X_test)

In [144]:
print('Explained Variance Score (R-squared): %.2f'
      % explained_variance_score(y_test, y_pred))
# The coefficient of determination: 1 is perfect prediction
print('Mean Absolute Error: %.2f'
      % mean_absolute_error(y_test, y_pred))

Explained Variance Score (R-squared): -305154320124414550605824.00
Mean Absolute Error: 2586201762840.33


## Review Model Outcomes — Iterate over additional models as needed.

In [145]:
#Intercept value from the linear model
lm.intercept_

651009409326.9098

**<font color='teal'> Print the intercept value from the linear model. </font>**

In [146]:
# Sort and print coeff
coeff_df = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
print(coeff_df.sort_values(by=['Coefficient'], ascending=False))

                    Coefficient
New York           9.992233e+12
Michigan           9.293273e+12
California         7.391708e+12
Pennsylvania       7.391708e+12
Wisconsin          6.968688e+12
New Hampshire      6.968688e+12
Colorado           6.744622e+12
Vermont            6.511043e+12
Minnesota          6.266890e+12
Montana            6.010872e+12
Massachusetts      6.010872e+12
Idaho              5.741403e+12
Washington         5.741403e+12
Wyoming            5.153575e+12
Utah               5.153575e+12
Maine              5.153575e+12
New Mexico         4.829268e+12
Oregon             4.829268e+12
Ohio               4.095873e+12
Connecticut        4.095873e+12
North Carolina     4.095873e+12
Virginia           3.669882e+12
West Virginia      3.669882e+12
Illinois           3.183762e+12
Iowa               3.183762e+12
Indiana            2.604056e+12
Arizona            2.604056e+12
New Jersey         2.604056e+12
Nevada             2.604056e+12
Alaska             2.604056e+12
Tennesse

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

In [147]:
# Pass X_train & X_test as DataFrame and add columns names
X_train_df = pd.DataFrame(X_train, columns=X.columns)
X_test_df = pd.DataFrame(X_test, columns=X.columns)


# Drop the states columns
X_train = X_train_df.drop(X_train_df.iloc[:,23:], axis = 1, inplace = False)
X_test = X_test_df.drop(X_test_df.iloc[:,23:], axis = 1, inplace = False)

The result as not as good as model one as the location of the resort plays a big importance on the price.

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [148]:
#model2
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X_train,y_train)

# Create a prediction y_pred
y_pred2 = lm2.predict(X_test)

print('Explained Variance Score (R-squared): %.2f'
      % explained_variance_score(y_test, y_pred2))
# The coefficient of determination: 1 is perfect prediction
print('Mean Absolute Error: %.2f'
      % mean_absolute_error(y_test, y_pred2))

Explained Variance Score (R-squared): 0.75
Mean Absolute Error: 6.30


In [149]:
# Sort and print coeff
coeff_df = pd.DataFrame(abs(lm2.coef_), X_test.columns, columns=['Coefficient'])
print(coeff_df.sort_values(by=['Coefficient'], ascending=False))

                   Coefficient
AdultWeekday         13.059204
quad                  1.900765
summit_elev           1.673705
kmean_cluster         1.634898
SkiableTerrain_ac     1.389582
vertical_drop         1.319085
trams                 1.272987
daysOpenLastYear      1.228319
projectedDaysOpen     1.193344
NightSkiing_ac        1.068156
total_chairs          0.861155
averageSnowfall       0.861034
fastQuads             0.843511
triple                0.785456
LongestRun_mi         0.605727
surface               0.544277
double                0.352227
fastSixes             0.332506
fastEight             0.315490
Snow Making_ac        0.217392
yearsOpen             0.215653
Runs                  0.040949
TerrainParks          0.018366


**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [150]:
# summit_elev was removed in step3 when checking high correlation between 2 features.
# Remove base_elev
X_train = X_train.drop(['summit_elev'], axis = 1)
X_test = X_test.drop(['summit_elev'], axis = 1)


#model3 
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X_train,y_train)

# Create a prediction y_pred
y_pred3 = lm3.predict(X_test)

print('Explained Variance Score (R-squared): %.2f'
      % explained_variance_score(y_test, y_pred3))
# The coefficient of determination: 1 is perfect prediction
print('Mean Absolute Error: %.2f'
      % mean_absolute_error(y_test, y_pred3))

Explained Variance Score (R-squared): 0.73
Mean Absolute Error: 6.61


In [151]:
# Sort and print coeff
coeff_df = pd.DataFrame(abs(lm3.coef_), X_test.columns, columns=['Coefficient'])
print(coeff_df.sort_values(by=['Coefficient'], ascending=False))

                   Coefficient
AdultWeekday         13.006530
quad                  1.939247
SkiableTerrain_ac     1.843976
averageSnowfall       1.464031
NightSkiing_ac        1.408704
kmean_cluster         1.343559
trams                 1.060264
daysOpenLastYear      0.987406
projectedDaysOpen     0.986666
vertical_drop         0.881814
total_chairs          0.850489
LongestRun_mi         0.818587
fastQuads             0.653341
triple                0.597186
Snow Making_ac        0.481213
fastSixes             0.417003
fastEight             0.371374
double                0.362983
surface               0.345001
yearsOpen             0.289082
Runs                  0.228152
TerrainParks          0.178652


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -305154320124414550605824 |2586201762840  |-|
| Model 2. | 0.75|6.3 |'state'|
| Model 3. | 0.73 |6.61 |'state','summit_elev','base_elev'|

Model Selection:

I will choose model 2 because the performance are slightly better and even if we can not change the summit elevation of our resort, model 2 will give us a better idea of the price range for resort of the same standings.